java/core/src/java/org/apache/orc/impl/DictionaryUtils.java

```diff
@@ -0,0 +1,86 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.orc.impl;
+
+import java.io.IOException;
+import java.io.OutputStream;
+
+import org.apache.hadoop.io.Text;
+
+
+public class DictionaryUtils {
+  private DictionaryUtils() {
+    // Utility class does nothing in constructor
+  }
+
+  public static void getTextInternal(Text result, int position, DynamicIntArray keyOffsets, DynamicByteArray byteArray) {
+    int offset = keyOffsets.get(position);
+    int length;
+    if (position + 1 == keyOffsets.size()) {
+      length = byteArray.size() - offset;
+    } else {
+      length = keyOffsets.get(position + 1) - offset;
+    }
+    byteArray.setText(result, offset, length);
+  }
+
+  static class VisitorContextImpl implements Dictionary.VisitorContext {
```

Comment:
I'd think that a common class (named DictionaryStorage?) that implemented VisitorContext and had the byteArray and keyOffsets would make sense. Each of the dictionaries could have a field that held the reference.

Reply:
Sorry, I left off the context of why. We could pull the common code into a single class between the red-black tree and hash table.